adapter/src/hermes/adapter/stdio/stdio.cc

```diff
@@ -0,0 +1,993 @@
+#include <limits.h>
+
+#include <hermes/adapter/interceptor.cc>
+#include <hermes/adapter/stdio/mapper/balanced_mapper.cc>
+#include <hermes/adapter/stdio/metadata_manager.cc>
+
+using hermes::adapter::stdio::AdapterStat;
+using hermes::adapter::stdio::FileID;
+using hermes::adapter::stdio::FileStruct;
+using hermes::adapter::stdio::MapperFactory;
+using hermes::adapter::stdio::MetadataManager;
+
+namespace hapi = hermes::api;
+namespace fs = std::experimental::filesystem;
+
+FILE *simple_open(FILE *ret, const std::string &path_str, const char *mode) {
+  LOG(INFO) << "Open file for filename " << path_str << " in mode " << mode
+            << std::endl;
+  auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+  if (!ret) {
+    return ret;
+  } else {
+    auto existing = mdm->Find(ret);
+    if (!existing.second) {
+      LOG(INFO) << "File not opened before by adapter" << std::endl;
+      struct stat st;
+      int fd = fileno(ret);
+      int status = fstat(fd, &st);
+      if (status == 0) {
+        AdapterStat stat(st);
+        stat.ref_count = 1;
+        struct timespec ts;
+        timespec_get(&ts, TIME_UTC);
+        stat.st_atim = ts;
+        stat.st_mtim = ts;
+        stat.st_ctim = ts;
+        if (strcmp(mode, "a") == 0 || strcmp(mode, "a+") == 0) {
+          /* FIXME: get current size of bucket from Hermes*/
+          stat.st_ptr = stat.st_size;
+        }
+        /* FIXME(hari) check if this initialization is correct. */
+        mdm->InitializeHermes();
+        hapi::Context ctx;
+        /* TODO(hari) how to pass to hermes to make a private bucket
+         * also add how to handle existing buckets of same name */
+        stat.st_bkid =
+            std::make_shared<hapi::Bucket>(path_str, mdm->GetHermes(), ctx);
+        mdm->Create(ret, stat);
+      } else {
+        // TODO(hari): @error_handling invalid fh.
+        ret = nullptr;
+      }
+    } else {
+      LOG(INFO) << "File opened before by adapter" << std::endl;
+      existing.first.ref_count++;
+      struct timespec ts;
+      timespec_get(&ts, TIME_UTC);
+      existing.first.st_atim = ts;
+      existing.first.st_ctim = ts;
+      mdm->Update(ret, existing.first);
+    }
+  }
+  return ret;
+}
+
+FILE *open_internal(const std::string &path_str, const char *mode) {
+  FILE *ret;
+  MAP_OR_FAIL(fopen);
+  ret = real_fopen_(path_str.c_str(), mode);
+  ret = simple_open(ret, path_str.c_str(), mode);
+  return ret;
+}
+
+FILE *reopen_internal(const std::string &path_str, const char *mode,
+                      FILE *stream) {
+  FILE *ret;
+  MAP_OR_FAIL(freopen);
+  auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+  ret = real_freopen_(path_str.c_str(), mode, stream);
+  if (!ret) {
+    return ret;
+  } else {
+    auto existing = mdm->Find(ret);
+    if (!existing.second) {
+      LOG(INFO) << "Reopen file for filename " << path_str << " in mode "
+                << mode << std::endl;
```

Comment:
Shouldn't this log message be outside of the current `if` block?

Reply:
Yap fixed.